interfaces/apparmor: load multiple profiles in a single batch

stolowski opened the pull request:
Per https://forum.snapcraft.io/t/apparmor-parser-performance-on-pi3/12909/2, this is a proposal of optimization of apparmor backend to load multiple profiles with a single invocation of `apparmor_parser`, and it's main use case is the regeneration of profiles on snapd startup (e.g. if system key changes).

This PR introduces an optional `SetupMany` security backend call that takes multiple snaps and sets them up in one go, and implements it for AppArmor backend.  Due to the possibility of a single profile failing the entire batch, `SetupMany` runs the batch first and if anything fails, it falls back to the one-by-one approach where all snaps are processed again individually, errors are recorded, but loading of profiles is not interrupted (I consider such error cases exeptional, and probably a sign of deeper problems if they happen).

To reduce the load on initial boot, apparmor invocation with SetupMany uses -jN flag (where N is less than the number of CPUs), as suggested by @jdstrand .

anonymouse64 commented on interfaces/apparmor/backend.go:
```diff
@@ -312,17 +312,11 @@ func profileIsRemovableOnCoreSetup(fn string) bool {
-// This method should be called after changing plug, slots, connections between
-// them or application present in the snap.
-func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions, repo *interfaces.Repository, tm timings.Measurer) error {
+func (b *Backend) prepareProfiles(snapInfo *snap.Info, opts interfaces.ConfinementOptions, repo *interfaces.Repository) (changedPaths, unchangedPaths, removedPaths []string, err error) {
```
could the return type be simplified here to use a struct like:

```go
type profilePathsResults struct {
    changed []string
    unchanged []string
    removed []string
}
```

that would make it a bit easier to read this and you wouldn't have as many `return nil, nil, nil, err` which looks a bit unwieldy IMHO

anonymouse64 commented on interfaces/apparmor/backend.go:
```diff
@@ -391,32 +389,44 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
-		errReloadChanged = loadProfiles(pathnames, cache, skipReadCache)
-	})
-
 	// Load all unchanged profiles anyway. This ensures those are correct in
```
is this comment still accurate? AFAICT we aren't loading any profiles in this function anymore

anonymouse64 commented on interfaces/apparmor/backend.go:
```diff
@@ -391,32 +389,44 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
+		return err
+	}
+
+	var errReloadChanged error
```
the out-of-date comment I mention above might need to be put here?

anonymouse64 commented on interfaces/apparmor/backend.go:
```diff
@@ -426,6 +436,59 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
+		timings.Run(tm, "load-profiles-many[unchanged]", fmt.Sprintf("load unchanged security profiles %d snaps", len(snaps)), func(nesttm timings.Measurer) {
+			errReloadOther = loadProfiles(allUnchangedPaths, dirs.AppArmorCacheDir, conserveCPU)
+		})
+		errUnload := unloadProfiles(allRemovedPaths, dirs.AppArmorCacheDir)
```
is there a reason we don't also measure how long it takes to run `unloadProfiles`? is it because it is always very fast?

jdstrand replied:
> is there a reason we don't also measure how long it takes to run `unloadProfiles`? is it because it is always very fast?

Perhaps (idk why it isn't) but can say at this time unloadProfiles only consists of unlinking files, not unloading from the kernel (see interfaces/apparmor/apparmor.go:unloadProfiles() for details).

stolowski replied:
Yes, that's exactly the reason, it's simply calling `UnlinkMany`, nothing to measure.

anonymouse64 commented on overlord/ifacestate/helpers_test.go:
```diff
@@ -417,6 +418,126 @@ apps:
+	mockSnaps(c, st)
+
+	// Pretend that security profiles are out of date and mock the
+	// function that writes the new system key with one always panics.
```
where does the function panic here?

stolowski replied:
Thanks, tweaked all the comments you pointed out.

anonymouse64 commented on interfaces/apparmor/apparmor.go:
```diff
@@ -70,6 +75,13 @@ func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
+	if flags&conserveCPU != 0 {
+		cpus := runtime.NumCPU()
+		if cpus > 1 {
+			// spare one CPU
```
it might be a good idea to explain why we spare one CPU in a doc-comment I think

pedronis replied:
why the choice of sparing one? we might usually be booting maybe at this point, there is a lot going on potentially? should we use half the cpus instead?

stolowski replied:
I went ahead with @jdstrand suggestion of sparing two CPUs for the time being, I'm open to changing this. Also moved it to a helper.

anonymouse64 commented on interfaces/apparmor/backend.go:
```diff
@@ -426,6 +436,59 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
 	return errUnload
 }
 
+func (b *Backend) SetupMany(snaps []*snap.Info, confinement func(snapName string) interfaces.ConfinementOptions, repo *interfaces.Repository, tm timings.Measurer) []error {
```
I think we should have a similar doc-comment as `Setup` on `SetupMany`

pedronis commented on interfaces/apparmor/backend.go:
```diff
@@ -426,6 +436,59 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
+
+	var errors []error
+	// if an error was encountered when processing all profiles at once, re-try them one by one
+	if fallback {
```
I'm definitely interested in @jdstrand input whether there might be problems with this approach

jdstrand replied:
> I'm definitely interested in @jdstrand input whether there might be problems with this approach

Actually, I really like the approach and agree with @stolowski that we shouldn't ever hit this case except in the face of snapd bugs or FS corruption. The policy should always compile and if it doesn't, that is an important bug to fix (this is why we have the interfaces-many* tests). Compiling in parallel is in general a good thing and this is a nice way to load everything and report back on issues. Plus, anything that did compile will already have the updated profiles on disk, with the regenerated cache and so the extra apparmor_parser calls are going to be essentially as fast as we can fork/exec, do two stat()s and reading the cache file off the disk for each profile. I also verified back to xenial that apparmor_parser will load everything it can and not bail on the first error. Considering how seldom we should ever fallback, this seems reasonable.

That said, I'd like to see a test case that corrupts an apparmor profile and demonstrates that the fallback code works as expected and shows the error for the bad profile. This will be tricky because corrupting the profile after snapd writes it be before apparmor_parser parses it will mean that the fallback will rewrite it to its pre-corrupted state. Creative parser mocking instead is possible.

stolowski replied:
Thank you! I've added unit tests covering the fallback scenario, hower a real life spread test is indeed tricky; I'll see if I can come up with something.

stolowski replied:
Spread test available now.

pedronis commented on overlord/ifacestate/helpers.go:
```diff
@@ -138,32 +138,55 @@ func (m *InterfaceManager) regenerateAllSecurityProfiles(tm timings.Measurer) er
-			timings.Run(tm, "setup-security-backend", fmt.Sprintf("setup security backend %q for snap %q", backend.Name(), snapInfo.InstanceName()), func(nesttm timings.Measurer) {
-				if err := backend.Setup(snapInfo, opts, m.repo, nesttm); err != nil {
+		// use .SetupMany() if implemented by the backend, otherwise fall back to .Setup()
+		if setupManyInterface, ok := backend.(interfaces.SecurityBackendSetupMany); ok {
```
It might make sense to make this a problem of interfaces pkg though.  Have an interfaces.SetupMany helper there that calls SetupMany or does a loops depending on the interface impl.

stolowski replied:
Done.

jdstrand commented on interfaces/apparmor/backend.go:
```diff
@@ -426,6 +436,59 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
+		if errUnload != nil {
+			logger.Noticef("failed to unload profiles: %s", errUnload)
+			fallback = true
+		}
```
Considering my comment regarding unload only unlinking, errUnload doesn't strictly have to trigger the fallback code. However, we want to get to the point of unloading once snap processes are gone, so doing something else for this case would be prematurely optimizing. Ie, leaving it here is good. :)

stolowski replied:
Indeed, I'd keep it that way for consistency.

stolowski commented:
This is ready for further reviews; I've addressed all the comments with the exception of a spread test that excercises batch/fallback scenario. I'll try to come up with something (potentially as a followup).

stolowski commented:
Added a spread test for this change.

jdstrand commented on tests/main/apparmor-batch-reload/task.yaml:
```diff
@@ -0,0 +1,65 @@
+    echo "Checking that apparmor profile cache has been populated"
+    # note, the list is not exhaustive but is enough for the test and the journal check below
+    for p in snap-update-ns.test-snapd-hello-multi-arch snap.test-snapd-hello-multi-arch.hello-i386 snap-update-ns.test-snapd-tools snap.test-snapd-tools.block snap.test-snapd-tools.cat; do
+        retry-tool -n 5 test -f $AA_CACHE/$p
```
fyi, 5 seconds is not necessarily a long time on, for example, an ARM device. Since this is a subset of profiles, I worry that this approach may be racy since apparmor_parser will parse these in the order given and it is possible there could be a few profiles prior to one of the ones in the list that will take longer and push this to fail. I'm not sure how you want to reorganize this, but you could wait on ps output to no longer have any apparmor_parsers and then perform your checks (though that doesn't work for the fallback case where the parsers are being called one after the other). You could calculate the total time to wait on that as something terrible like (10*number of profiles in AA_PROFILES).

stolowski replied:
Thank you, that's a good observation, I changed it to retry a single test -f check over multiple profiles at once, with a more generous number of retries.

stolowski replied:
I actually re-worked it slightly to use `find` to discover profiles in AA cache, and then diff them with expected set. This should make it happy with "cache as a forest" in newer apparmor versions (thanks for pointing this AA change out!).

jdstrand commented on tests/main/apparmor-batch-reload/task.yaml:
```diff
@@ -0,0 +1,65 @@
+    echo $log | MATCH "FAIL ON: .*/snap-update-ns.test-snapd-tools"
+    echo $log | MATCH "FAIL ON: .*/snap.test-snapd-tools.block"
+    echo $log | MATCH "FAIL ON: .*/snap.test-snapd-tools.cat"
+    
```
Missing newline.

stolowski commented on interfaces/apparmor/apparmor.go:
```diff
@@ -70,6 +93,9 @@ func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
 	// Use no-expr-simplify since expr-simplify is actually slower on armhf (LP: #1383858)
 	args := []string{"--replace", "--write-cache", "-O", "no-expr-simplify", fmt.Sprintf("--cache-loc=%s", cacheDir)}
+	if flags&conserveCPU != 0 {
+		args = maybeSetNumberOfJobs(args)
```
I like this suggestions, thanks.

stolowski commented on interfaces/apparmor/backend_test.go:
```diff
@@ -369,6 +403,140 @@ func (s *backendSuite) TestUpdatingSnapToOneWithFewerHooks(c *C) {
+		snap2nsProfile := filepath.Join(dirs.SnapAppArmorDir, "snap-update-ns.some-snap")
+		snap2AAprofile := filepath.Join(dirs.SnapAppArmorDir, "snap.some-snap.someapp")
+
+		// We expect three calls to apparmor_parser due to the failure of batch run. First is the failed batch run, followed by succesfull fallback runs.
```
Added both the comment and a separate unit test.

stolowski commented:
This is ready for further reviews.

stolowski commented on tests/main/apparmor-batch-reload/task.yaml:
```diff
@@ -0,0 +1,95 @@
+
+    echo "Checking that apparmor profile cache has been populated"
+    # note, the list is not exhaustive but is enough for the test and the journal check below
+    expected_profiles=$(cat <<EOF
```
Indeed, that works too, thanks!

stolowski commented on tests/main/apparmor-batch-reload/bin/apparmor_parser.fake:
```diff
@@ -0,0 +1,43 @@
+            # preprocess is used by snapd to verify apparmor features
+            exit 0
+            ;;
+		--cache-loc=*)
```
Right, should be fixed now.

stolowski commented on tests/main/apparmor-batch-reload/bin/apparmor_parser.fake:
```diff
@@ -0,0 +1,43 @@
+
+# otherwise run real apparmor parser
+# shellcheck disable=SC2086
+/sbin/apparmor_parser.real $args
```
That's because the $@ is processed by the loop above (and modified via shift).

stolowski commented on tests/main/apparmor-batch-reload/task.yaml:
```diff
@@ -0,0 +1,92 @@
+    rm -f /tmp/apparmor-parser-calls.debug profiles.out
+
+debug: |
+    cat /tmp/apparmor-parser-calls.debug || true
```
In case the test fails early before the log gets created; anyway this is now entirely removed per the other comment related to potential symlink attack.

stolowski commented on tests/main/apparmor-batch-reload/bin/apparmor_parser.fake:
```diff
@@ -0,0 +1,43 @@
+debugfile=/tmp/apparmor-parser-calls.debug
+# shellcheck disable=SC2124
+args=$@
+echo "$@" >> $debugfile
```
Right. I couldn't think of a better place other than some other hardcoded path that's only writable by root. Since this log file is not super important for the test itself, I changed it to simply log do /dev/kmsg so if it fails, we still have a place to investigate.

stolowski commented on interfaces/apparmor/backend_test.go:
```diff
@@ -370,6 +431,171 @@ func (s *backendSuite) TestUpdatingSnapToOneWithFewerHooks(c *C) {
+}
+
+// helper for checking for apparmor parser calls: batch run followed by fallback runs
+func (s *backendSuite) checkSetupManyCallsWithFallback(c *C, cmd *testutil.MockCmd) {
```
This helper is shared between all "failing" scenarios, because they all test the same pattern: first aa parser call with multiple profiles fails, and is followed by calls to individual profiles. The calls are the same. The differences are in each individual test cases (the checks before/after calling this helper) and the errors that are returned  from `SetupMany()` and logged to the logger. I've added a few more comments to the test to hopefully make it more clear.

stolowski commented on tests/main/apparmor-batch-reload/task.yaml:
```diff
@@ -0,0 +1,92 @@
+    # remove all apparmor cached profiles so we can check they are recreated
+    rm -f $AA_CACHE/snap* $AA_CACHE/*/snap*
+
+    # "changed" variant requires source profiles on the disk to look different than the
```
Good point, I forgot to add anything besides a very short summary. Added "details" section to the test, and a few more comments inline.

stolowski commented on tests/main/apparmor-batch-reload/task.yaml:
```diff
@@ -0,0 +1,92 @@
+    # grep by "fail" to reduce log/debug noise in case of a match failure below
+    log="$(get_journalctl_log -a -u snapd|grep -i fail)"
+    echo "$log" | MATCH "failed to batch-reload $VARIANT profiles: cannot load apparmor profiles"
+    # we're matching a single long log line with non-deterministic order of profiles, therefore it's split
```
Added a comment.

pedronis commented on interfaces/helpers.go:
```diff
@@ -0,0 +1,64 @@
+	profilesOk := true
+	// use .SetupMany() if implemented by the backend, otherwise fall back to .Setup()
+	if setupManyInterface, ok := backend.(SecurityBackendSetupMany); ok {
+		timings.Run(tm, "setup-security-backend", fmt.Sprintf("setup security backend %q", backend.Name()), func(nesttm timings.Measurer) {
```
this should say "many" in the the label and summary

stolowski replied:
Couldn't find a perfect one - renamed to "setup-security-backend[many]" and "setup security backend %q for %d snaps"; what do you think?

pedronis commented on interfaces/helpers.go:
```diff
@@ -0,0 +1,64 @@
+
+// SetupMany generates profiles of snaps using either SetupMany() method of the security backend (if implemented), or Setup(). All errors are logged.
+// The return value indicates if all profiles were successfully generated.
+func SetupMany(repo *Repository, backend SecurityBackend, snaps []*snap.Info, confinementOpts func(snapName string) ConfinementOptions, tm timings.Measurer) bool {
```
not quite a blocker but I would have expected this to return an error/errors, not just a bool and do the logging one level up

pedronis commented on interfaces/apparmor/backend.go:
```diff
@@ -427,6 +444,67 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
+
+	if !fallback {
+		var errReloadChanged error
+		timings.Run(tm, "load-profiles-many[changed]", fmt.Sprintf("load changed security profiles of %d snaps", len(snaps)), func(nesttm timings.Measurer) {
```
load-profiles[changed-many]

pedronis commented on interfaces/apparmor/backend.go:
```diff
@@ -427,6 +444,67 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
+		})
+
+		var errReloadOther error
+		timings.Run(tm, "load-profiles-many[unchanged]", fmt.Sprintf("load unchanged security profiles %d snaps", len(snaps)), func(nesttm timings.Measurer) {
```
load-profiles[unchanged-many]

pedronis commented on interfaces/apparmor/backend.go:
```diff
@@ -427,6 +444,67 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
+		for _, snapInfo := range snaps {
+			opts := confinement(snapInfo.InstanceName())
+			if err := b.Setup(snapInfo, opts, repo, tm); err != nil {
+				errors = append(errors, fmt.Errorf("failed to setup profiles for snap %q: %s", snapInfo.InstanceName(), err))
```
s/failed to/cannot/
